### PR TITLE
Update for Swift 4.2  

### PR DESCRIPTION
**Pull request by mattt**
This PR improves support for Swift 4.2 with the following changes:

- Add support for keywords beginning with `#` (e.g. `#available`)
- Fix syntax for octal integer literals (e.g. `0o744`)
- Add support for multi-line string literals, which are delimited by three quotation marks (`"""`)

### CODE REVIEW

**Comment by mwaterfall** [comment]
🙏🏻 Would love this to be merged!

**Comment by pyrmont** [comment]
@mattt Sorry it's taken so long to get around to this one :( Would you be able to add some of these to the visual sample? Don't need to list all the keywords, just a couple of additions to show this working properly.

**Comment by mattt** [comment]
@pyrmont Sure thing! See 08b9917

**Comment by pyrmont** [comment]
This generally looks fine but I noticed that the `p` in the hexadecimal floating point isn't being picked up as a number. Shouldn't it be?

**Comment by mattt** [comment]
@pyrmont That wasn't in the original scope of this patch, but I just implemented it in 
bbec13e.

Tangentially: What's your process for checking / verifying the result of the visual sample?

File: lib/rouge/lexers/swift.rb [hunk 1]
```diff
@@ -73,8 +75,8 @@ class deinit enum convenience extension final func import init internal lazy let
         rule /'(\\.|.)'/, Str::Char
         rule /(\d+\*|\d*\.\d+)(e[+-]?[0-9]+)?/i, Num::Float
```
Comment (pyrmont) [inline lib/rouge/lexers/swift.rb]: This rule as written doesn't match floating-point numbers with groupings (you can see this in the visual sample). I had a quick play around and I think this works:

```ruby
rule /(\d+(?:_\d+)*\*|(?:\d+(?:_\d+)*)*\.\d+(?:_\d)*)(e[+-]?\d+(?:_\d)*)?/i, Num::Float
```

File: lib/rouge/lexers/swift.rb [hunk 2]
```diff
@@ -73,8 +75,8 @@ class deinit enum convenience extension final func import init internal lazy let
...
         rule /\d+e[+-]?[0-9]+/i, Num::Float
```
Comment (pyrmont) [inline lib/rouge/lexers/swift.rb]: It's not a big deal but for consistency, I'd suggest consistent use of `\d`:

```ruby
rule /\d+e[+-]?\d+/i, Num::Float
```

File: lib/rouge/lexers/swift.rb [hunk 3]
```diff
@@ -73,8 +75,8 @@ class deinit enum convenience extension final func import init internal lazy let
...
-        rule /0_?[0-7]+(?:_[0-7]+)*/, Num::Oct
-        rule /0x[0-9A-Fa-f]+(?:_[0-9A-Fa-f]+)*/, Num::Hex
+        rule /0o?[0-7]+(?:_[0-7]+)*/, Num::Oct
+        rule /0x[0-9A-Fa-f]+(?:_[0-9A-Fa-f]+)*((\.[0-9A-F]+(?:_[0-9A-F]+)*)?p\d+)?/, Num::Hex
```
Comment (pyrmont) [inline lib/rouge/lexers/swift.rb]: It looks like from what I can find online (and the visual sample itself) that there should be a `-` after the `p`. Something like this:

```ruby
rule /0x[0-9A-Fa-f]+(?:_[0-9A-Fa-f]+)*((\.[0-9A-F]+(?:_[0-9A-F]+)*)?p-\d+)?/, Num::Hex
```

**Comment by pyrmont** [comment]
@mattt wrote:

> Tangentially: What's your process for checking / verifying the result of the visual sample?

There probably should be a more mechanical (and, hence, reliable) process but honestly, at the moment, I just look at it with the naked eye and try to find mistakes. That's basically how I identified the problems with the floating point/hex regular expressions.

File: lib/rouge/lexers/swift.rb (hunk 3 above)
Comment (mattt) [reply lib/rouge/lexers/swift.rb]: You're right, this is missing an optional sign. This is fixed by 1384b71fac47c3fe0aa8c46b7a0f574afb644a49.

For reference, here's what I could find in the way of authoritative examples of valid and invalid hexadecimal floating-point literals: https://github.com/apple/swift/blob/2ee371b03a0fb0623188d564399681f89e57c36a/test/expr/expressions.swift#L414-L427

Comment (pyrmont) [reply lib/rouge/lexers/swift.rb]: Awesome! You figure they'd know :)

File: lib/rouge/lexers/swift.rb (hunk 2 above)
Comment (mattt) [reply lib/rouge/lexers/swift.rb]: I don't disagree, but I'd like to investigate further improvements separate to this PR.

**Comment by mattt** [comment]
> There probably should be a more mechanical (and, hence, reliable) process but honestly, at the moment, I just look at it with the naked eye and try to find mistakes. That's basically how I identified the problems with the floating point/hex regular expressions.

Oh, wow. I was expecting there to be some command that you could run to quickly generate an HTML preview of the current rules applied to the visual sample. Does that not exist? And if not, is this something you'd consider in a PR?

**Comment by pyrmont** [comment]
@mattt Oh, wait. I may have misunderstood your original question. You realise if you run `rackup` from the project root, it'll spin up a quick server at http://localhost:8000/ that will display the visual samples as webpages, right?

**Comment by pyrmont** [comment]
@mattt I hope you don't mind but I took the liberty of pushing a commit to your branch with the necessary fix. Hopefully that means we can get this merged :)

I also realised that the link I gave you before was wrong. I manually set my port to be 8000 but by default running `bundle exec rackup` will launch the server at http://localhost:9292/.

Hope that helps for any future PRs!

**Comment by pyrmont** [comment]
@mattt This has been successful merged. Thanks for making Rouge a better library :)

**Comment by mattt** [comment]
@pyrmont Thanks for your help and guidance through the process. Looking forward to sending more PRs in the future.

**Pull request merged**